h2o-core/src/main/java/water/rapids/SortCombine.java

```diff
@@ -0,0 +1,407 @@
+    _numRowsInResult = retSize;
+    
+    setPerNodeNumsToFetch();  // find out the number of rows to fetch from H2O nodes, number of rows to fetch per chunk
+    
+    if (_numRowsInResult > 0) createChunksInDKV();
+    tryComplete();
+  }
+
+  /**
+   * This method will find the number of rows per chunk to fetch per batch for a MSB
+   */
+  private void setPerNodeNumsToFetch() {
+    Vec anyVec = _leftSB._frame.anyVec();
+    int nbatch = _leftKO._order.length;
+    _numRowsPerCidx = new int[nbatch][anyVec.nChunks()];
+    for (int batchNum =0; batchNum < nbatch; batchNum++) {
+      int batchSize = _leftKO._order[batchNum].length;
+      for (int index=0; index < batchSize; index++) {
+        long globalRowNumber = _leftKO._order[batchNum][index];
+        int chkIdx = _leftSB._vec.elem2ChunkIdx(globalRowNumber);
+        _leftKO._perNodeNumRowsToFetch[_leftSB._chunkNode[chkIdx]]++;
+        _numRowsPerCidx[batchNum][chkIdx]++;  // number of rows to fetch per Cidx for a certain MSB
+      }
+    }
+  }
+
+  // Holder for Key & Order info
+  private static class KeyOrder {
+    public final transient long _batchSize;
+    private final transient byte _key[/*n2GB*/][/*i mod 2GB * _keySize*/];
+    private final transient long _order[/*n2GB*/][/*i mod 2GB * _keySize*/];
+    private final transient long _perNodeNumRowsToFetch[];
+
+    KeyOrder(SingleThreadRadixOrder.OXHeader sortedOXHeader) {
+      _batchSize = sortedOXHeader._batchSize;
+      final int nBatch = sortedOXHeader._nBatch;
+      _key = new byte[nBatch][];
+      _order = new long[nBatch][];
+      _perNodeNumRowsToFetch = new long[H2O.CLOUD.size()];
+    }
+
+    void initKeyOrder(int msb, boolean isLeft) {
+      for (int b = 0; b < _key.length; b++) {
+        Value v = DKV.get(SplitByMSBLocal.getSortedOXbatchKey(isLeft, msb, b));
+        SplitByMSBLocal.OXbatch ox = v.get(); //mem version (obtained from remote) of the Values gets turned into POJO version
+        v.freeMem(); //only keep the POJO version of the Value
+        _key[b] = ox._x;
+        _order[b] = ox._o;
+      }
+    }
+  }
+
+  private void chunksPopulatePerChunk(final long[][][] perNodeLeftRowsCidx, final long[][][] perNodeLeftIndices) {
+    int numBatch = _leftKO._order.length; // note that _order is ordered as nbatch/batchIndex
+
+    int[][] chkIndices = new int[numBatch][]; // store index for each batch per chunk
+    for (int nbatch=0; nbatch < numBatch; nbatch++) {
+      int sortedRowIndex = -1;
+      chkIndices[nbatch] = new int[_chunkNum];
+      int batchSize = _leftKO._order[nbatch].length;
+      for (int batchNum=0; batchNum < batchSize; batchNum++) {
+        sortedRowIndex++;
+        long row = _leftKO._order[nbatch][batchNum];
+        int chkIdx =  _leftSB._vec.elem2ChunkIdx(row); //binary search in espc
+        perNodeLeftRowsCidx[nbatch][chkIdx][chkIndices[nbatch][chkIdx]] = row;
+        perNodeLeftIndices[nbatch][chkIdx][chkIndices[nbatch][chkIdx]] = sortedRowIndex;
```

Comment:
This is the second time we go over the whole dataframe to figure out for each batch, and for each chunk, what the global row number of the row (perNodeLeftRowsCidx) should be and in what order (perNodeLeftIndices) should we put it back together.